### PR TITLE
Add Python 3 support

### LINKED ISSUE #138: Make Clipper compatible with Python 3
While running the tutorials, I realized that Clipper is not supporting Python 3. It will be necessary to update the python files to add support to Python 3 while also maintaining support for Python 2.

It will also be necessary to check the dependencies of the project. For example, [fabric](https://github.com/fabric/fabric) is only available for Python 2.

### PR DESCRIPTION
**Pull request by jwirjo**
Fixes #138. Added Python3 ClipperPy3TestsDockerfile, ran most integration_tests (run_unittests.sh) in python3 with same output as under python2, but 2 recurring errors.

### CODE REVIEW

**Comment by dcrankshaw** [comment]
@jwirjo Can you delete all the extra files you added in `bin/`? That will make this much easier to review.

**Comment by dcrankshaw** [comment]
Sorry meant to say in `clipper_admin/build`

File: clipper_admin/clipper_admin/deployers/deployer_utils.py [hunk 1]
```diff
@@ -42,6 +42,7 @@ def save_python_function(name, func):
     c.dump(func)
     serialized_prediction_function = s.getvalue()
 
+
```
Comment (dcrankshaw) [inline clipper_admin/clipper_admin/deployers/deployer_utils.py]: Revert this change

File: clipper_admin/clipper_admin/deployers/python.py [hunk 2]
```diff
@@ -21,8 +22,9 @@ def create_endpoint(
         slo_micros=3000000,
         labels=None,
         registry=None,
-        base_image="clipper/python-closure-container:{}".format(__version__),
+        base_image="default",
         num_replicas=1):
+
```
Comment (dcrankshaw) [inline clipper_admin/clipper_admin/deployers/python.py]: Delete extra line

File: clipper_admin/clipper_admin/deployers/python.py [hunk 3]
```diff
@@ -72,6 +74,7 @@ def create_endpoint(
         :py:meth:`clipper.ClipperConnection.set_num_replicas`.
     """
 
+
```
Comment (dcrankshaw) [inline clipper_admin/clipper_admin/deployers/python.py]: Delete extra line

File: clipper_admin/clipper_admin/docker/docker_container_manager.py [hunk 4]
```diff
@@ -265,8 +265,12 @@ def get_logs(self, logging_dir):
...
                 image=c.image.short_id, id=c.short_id)
             log_file = os.path.join(logging_dir, log_file_name)
-            with open(log_file, "w") as lf:
-                lf.write(c.logs(stdout=True, stderr=True))
+            try:
+                with open(log_file, "w") as lf:
+                    lf.write(c.logs(stdout=True, stderr=True))
+            except TypeError:
+                with open(log_file, "wb") as lf:
+                    lf.write(c.logs(stdout=True, stderr=True))
```
Comment (dcrankshaw) [inline clipper_admin/clipper_admin/docker/docker_container_manager.py]: Is this a Python 2 vs 3 thing? If so, instead of trying one way and catching an exception, let's explicitly check for `sys.version` like you do in the python deployer.

File: containers/python/rpc.py [hunk 5]
```diff
@@ -332,7 +332,10 @@ def run(self, metric_conn):
                 sys.stderr.flush()
 
     def send_container_metadata(self, socket):
-        socket.send("", zmq.SNDMORE)
+        try:
+            socket.send("", zmq.SNDMORE)
+        except:
+            socket.send("".encode('utf-8'), zmq.SNDMORE)
```
Comment (dcrankshaw) [inline containers/python/rpc.py]: Is this a Python 2 vs 3 thing? If so, instead of trying one way and catching an exception, let's explicitly check for sys.version like you do in the python deployer and then just use the right version. Throwing and catching an exception can be expensive and the code as-is has unclear semantics.

File: containers/python/rpc.py [hunk 6]
```diff
@@ -343,7 +346,10 @@ def send_container_metadata(self, socket):
         sys.stderr.flush()
 
     def send_heartbeat(self, socket):
-        socket.send("", zmq.SNDMORE)
+        try:
+            socket.send("", zmq.SNDMORE)
+        except TypeError:
+            socket.send_string("", zmq.SNDMORE)
```
Comment (dcrankshaw) [inline containers/python/rpc.py]: Same comment on checking for `sys.version`

File: containers/python/rpc.py [hunk 7]
```diff
@@ -400,7 +406,10 @@ def add_output(self, output):
         ----------
         output : string
         """
-        output = unicode(output, "utf-8").encode("utf-8")
+        if not isinstance(output, str):
+            output = unicode(output, "utf-8").encode("utf-8")      
+        else:
+            output = output.encode('utf-8')
```
Comment (dcrankshaw) [inline containers/python/rpc.py]: Same comment on sys.version

File: containers/python/rpc.py [hunk 8]
```diff
@@ -410,7 +419,10 @@ def add_output(self, output):
         self.string_content_end_position += output_len
 
     def send(self, socket, event_history):
-        socket.send("", flags=zmq.SNDMORE)
+        try:
+            socket.send("", flags=zmq.SNDMORE)
+        except TypeError:
+            socket.send_string("", flags=zmq.SNDMORE)
```
Comment (dcrankshaw) [inline containers/python/rpc.py]: same comment on sys.version

File: containers/python/send_queries.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+                  headers=headers,
+                  data=json.dumps({"input": list(features)})).json()
+    if not response["default"]:
+        if np.allclose(float(response["output"]), np.sum(features)):
+            print("Success")
+        else:
+            print("Received non-default response but value was incorrect")
+    else:
+       print("Received default response")
+    time.sleep(2)
```
Comment (dcrankshaw) [inline containers/python/send_queries.py]: This just looks like a test file. Delete this.

File: containers/python/sum_container.py [hunk 10]
```diff
@@ -22,6 +22,7 @@ def predict_bytes(self, inputs):
         return [str(sum(item)) for item in inputs]
 
     def predict_strings(self, inputs):
+        print(inputs)
```
Comment (dcrankshaw) [inline containers/python/sum_container.py]: revert

File: dockerfiles/ClipperPy3TestsDockerfile [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+RUN pip install requests subprocess32 scikit-learn numpy pyyaml docker cloudpickle==0.5.2
+
+COPY ./ /clipper
+
+# Patch Clipper
+RUN cd /clipper/src/libs/spdlog \
+    && git apply ../patches/make_spdlog_compile_linux.patch
+
+
+ENTRYPOINT ["/clipper/bin/ci_checks.sh"]
```
Comment (dcrankshaw) [inline dockerfiles/ClipperPy3TestsDockerfile]: We'll want to modify this entrypoint command to just run the Python 3-compatible tests, not re-run everything

File: dockerfiles/Py3ClosureContainerDockerfile.txt [hunk 12]
```diff
@@ -0,0 +1,13 @@
+ARG CODE_VERSION
```
Comment (dcrankshaw) [inline dockerfiles/Py3ClosureContainerDockerfile.txt]: The Dockerfile file name should not end in `.txt`

File: dockerfiles/Py3RPCDockerfile.txt [hunk 13]
```diff
@@ -0,0 +1,20 @@
+# This ARG isn't used but prevents warnings in the build script
```
Comment (dcrankshaw) [inline dockerfiles/Py3RPCDockerfile.txt]: The Dockerfile file name should not end in `.txt`

File: integration-tests/clipper_admin_tests.py [hunk 14]
```diff
@@ -472,30 +472,30 @@ def predict_func(inputs):
...
+    #'test_set_num_replicas_for_external_model_fails',
+    #'test_model_version_sets_correctly',
+    #'test_get_logs_creates_log_files',
+    #'test_inspect_instance_returns_json_dict',
+    #'test_model_deploys_successfully',
+    #'test_set_num_replicas_for_deployed_model_succeeds',
+    #'test_remove_inactive_containers_succeeds',
+    #'test_stop_models',
+    #'test_python_closure_deploys_successfully',
     'test_register_py_endpoint',
```
Comment (dcrankshaw) [inline integration-tests/clipper_admin_tests.py]: Why are these all commented out?

**Review by dcrankshaw — CHANGES_REQUESTED** [review]
This looks pretty good. I'm going to play around with this on my machine as well, but I added a first round of comments.

File: bin/build_docker_images.sh [hunk 15]
```diff
@@ -245,16 +245,20 @@ build_images () {
     create_image query_frontend QueryFrontendDockerfile $public
     create_image management_frontend ManagementFrontendDockerfile $public
     create_image unittests ClipperTestsDockerfile  $private
+    create_image unittests ClipperPy3TestsDockerfile  $private
```
Comment (dcrankshaw) [inline bin/build_docker_images.sh]: Change the name of the Py3 unittests image to `py3tests`

File: .pypirc [hunk 16]
```diff
@@ -0,0 +1,10 @@
+[distutils]
+index-servers=
+    pypi
+    test
+
+[test]
+repository = https://test.pypi.org/legacy/
+username = jwirjo
+password = clippertest
```
Comment (dcrankshaw) [inline .pypirc]: You definitely don't want to commit this!

File: bin/run_unittests.sh [hunk 17]
```diff
@@ -197,7 +198,7 @@ case $args in
     -r | --rpc-container )  set_test_environment
                             run_rpc_container_tests
                             ;;
-    -i | --integration_tests )  set_test_environment
+    -i | --integration_tests )  #set_test_environment
```
Comment (dcrankshaw) [inline bin/run_unittests.sh]: revert this change

File: dockerfiles/Py3ClosureContainerDockerfile [hunk 18]
```diff
@@ -0,0 +1,13 @@
+ARG CODE_VERSION
+FROM clipper/py3-rpc:${CODE_VERSION}
+
+COPY clipper_admin/clipper_admin/python3_container_conda_deps.txt /lib/
+RUN conda install -y --file /lib/python3_container_conda_deps.txt
+RUN conda install -c anaconda3 cloudpickle=0.5.2
```
Comment (dcrankshaw) [inline dockerfiles/Py3ClosureContainerDockerfile]: Change lines 4-5 to:

```
RUN conda config --set ssl_verify no \
    && conda install -c anaconda3 cloudpickle=0.5.2 \
    && conda install -y --file /lib/python3_container_conda_deps.txt
```
This should fix the current test failure (see #426)

**Review by dcrankshaw — CHANGES_REQUESTED** [review]
This is super close! I added a comment that should help get the tests passing, and then this will be good to go.

File: bin/run_unittests.sh [hunk 19]
```diff
@@ -140,6 +140,7 @@ function run_integration_tests {
   python ../integration-tests/deploy_pyspark_models.py
   python ../integration-tests/deploy_pyspark_pipeline_models.py
   python ../integration-tests/deploy_pyspark_sparkml_models.py
+  python ../integration-tests/deploy_pytorch_models.py
```
Comment (dcrankshaw) [inline bin/run_unittests.sh]: Since this test seems to be failing, you can revert it and I'll add it in a separate PR.

**Comment by dcrankshaw** [comment]
jenkins test this please

**Comment by dcrankshaw** [comment]
jenkins test this please

**Comment by dcrankshaw** [comment]
jenkins test this please

**Comment by marcusklaas** [comment]
Yo Dan, landing this PR would be huge! How do things look at this time?

**Comment by dcrankshaw** [comment]
@marcusklaas This is basically ready to go. We're in the midst of debugging some issues with our CI cluster that's caused a bit of a build up in getting PRs merged. Should have that resolved by tomorrow, then this and several other PRs can get merged.

**Review by dcrankshaw — APPROVED** [review]
This is good to go once the tests pass

**Comment by dcrankshaw** [comment]
jenkins test this please

**Comment by dcrankshaw** [comment]
jenkins test this please

**Comment by dcrankshaw** [comment]
jenkins test this please

**Comment by dcrankshaw** [comment]
jenkins test this please

**Comment by dcrankshaw** [comment]
jenkins test this please

**Comment by dcrankshaw** [comment]
jenkins test this please

**Comment by dcrankshaw** [comment]
@jwirjo @rohsuresh Any progress on this?

**Comment by dcrankshaw** [comment]
jenkins test this please

**Comment by dcrankshaw** [comment]
jenkins test this please
